pkg/metrics/metrics.go

```diff
@@ -0,0 +1,218 @@
+				Buckets:   prometheus.LinearBuckets(.000, .05, 30),
+			},
+			[]string{"authenticated", "code", "remote_address", "user"},
+		)
+	)
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(clientRequests)
+	registry.MustRegister(clientDuration)
+	registry.MustRegister(serverRequests)
+	registry.MustRegister(serverDuration)
+	registry.MustRegister(oidcAuthCount)
+	registry.MustRegister(tokenReviewDuration)
+
+	return &Metrics{
+		registry: registry,
+
+		clientRequests: clientRequests,
+		clientDuration: clientDuration,
+
+		serverRequests: serverRequests,
+		serverDuration: serverDuration,
+
+		oidcAuthCount:       oidcAuthCount,
+		tokenReviewDuration: tokenReviewDuration,
+	}
+}
+
+// Start will register the Prometheus metrics, and start the Prometheus server
+func (m *Metrics) Start(listenAddress string) error {
+	router := http.NewServeMux()
+	router.Handle("/metrics", promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{}))
+
+	ln, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return err
+	}
+
+	m.Server = &http.Server{
+		Addr:           ln.Addr().String(),
+		ReadTimeout:    8 * time.Second,
+		WriteTimeout:   8 * time.Second,
+		MaxHeaderBytes: 1 << 15, // 1 MiB
+		Handler:        router,
+	}
+
+	go func() {
+		klog.Infof("serving metrics on %s/metrics", ln.Addr())
+
+		if err := m.Serve(ln); err != nil {
+			klog.Errorf("failed to serve prometheus metrics: %s", err)
+			return
+		}
+	}()
+
+	return nil
+}
+
+func (m *Metrics) Shutdown() error {
+	// If metrics server is not started than exit early
+	if m.Server == nil {
+		return nil
+	}
+
+	klog.Info("shutting down Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	if err := m.Server.Shutdown(ctx); err != nil {
+		return fmt.Errorf("prometheus metrics server shutdown failed: %s", err)
+	}
+
+	klog.Info("prometheus metrics server gracefully stopped")
+
+	return nil
+}
+
+func (m *Metrics) ObserveClient(code int, path, remoteAddress string, duration time.Duration) {
+	m.clientRequests.With(prometheus.Labels{
+		"code":           strconv.Itoa(code),
+		"path":           path,
+		"remote_address": remoteAddress,
+	}).Inc()
+
+	m.clientDuration.With(prometheus.Labels{
+		"remote_address": remoteAddress,
+	}).Observe(duration.Seconds())
+}
+
+func (m *Metrics) ObserveServer(code int, path, remoteAddress string, duration time.Duration) {
+	m.serverRequests.With(prometheus.Labels{
+		"code":           strconv.Itoa(code),
+		"path":           path,
+		"remote_address": remoteAddress,
+	}).Inc()
+
+	m.serverDuration.With(prometheus.Labels{
+		"remote_address": remoteAddress,
+	}).Observe(duration.Seconds())
+}
+
+func (m *Metrics) IncrementOIDCAuthCount(authenticated bool, remoteAddress, user string) {
+	m.oidcAuthCount.With(prometheus.Labels{
+		"authenticated":  boolToIntString(authenticated),
+		"remote_address": remoteAddress,
+		"user":           user,
```

Comment:
This will be a different user for every user on the cluster? That is probably fine for a label.

Reply:
Indeed, that comes from the auth when validating the token

Reply:
Yeah, so we would expect some fairly low number of users in most cases (hundreds at the most), so I think that would be ok for a label.